homeassistant/components/climate/__init__.py

```diff
@@ -556,6 +558,19 @@ async def async_turn_off(self) -> None:
         if HVACMode.OFF in self.hvac_modes:
             await self.async_set_hvac_mode(HVACMode.OFF)
 
+    async def async_toggle(self) -> None:
+        """Toggle the entity."""
+        if hasattr(self, "toggle"):
+            await self.hass.async_add_executor_job(self.toggle)
+            return
```

Comment:
We could also have a base implementation in the base class, maybe that makes it more clear it can be extended?

```suggestion
    def toggle(self) -> None:
        """Toggle the entity."""
        raise NotImplementedError

    async def async_toggle(self) -> None:
        """Toggle the entity."""
        if self.toggle is not ClimateEntity.toggle:
            await self.hass.async_add_executor_job(self.toggle)
            return
```

Reply:
I think that in general we should promote the async pattern more and consider the sync versions more of a legacy thing, so I would be rather against this change.
Also the current async_turn_on and async_turn_off use the same pattern, so if this was changed, those two probably should as well.

Reply:
> I think that in general we should promote the async pattern more and consider the sync versions more of a legacy thing, so I would be rather against this change.

That's not really the case. The sync version is needed to integrate with 3rd party libraries which do not themselves use `asyncio`

Reply:
An alternative is to do the same implementation as `ToggleEntity`:

```py
    @final
    def toggle(self, **kwargs: Any) -> None:
        """Toggle the entity.

        This method will never be called by Home Assistant and should not be implemented
        by integrations.
        """

    @final
    async def async_toggle(self, **kwargs: Any) -> None:
        """Toggle the entity.

        This method should not be implemented by integrations, it's enough to
        implement async_turn_on + async_turn_off or turn_on + turn_off.
        """
        if HVACMode.OFF in self.hvac_modes:
            if self.hvac_mode == HVACMode.OFF:
                await self.async_turn_on()
            else:
                await self.async_turn_off()
```